### PR TITLE
Add RequestFactory and RequestTransformer.

### PR DESCRIPTION
**Pull request by markstory**
The ServerRequestFactory is responsible for:
- Building a request from the SAPI super globals.
- Extracting the base and webroot directories for backwards  compatibility with the CakeRequest.
- Updating the request path to reflect only the application's 'virtual path'

The RequestTransformer handles:
- Converting a PSR7 request object into the equivalent Cake\Network\Http  request.
- Ensuring that the required routing parameters are set even if the PSR7  request is missing the 'params' attribute.

In order to shim the backwards compatibility, we'll use 3 attributes which keep track of CakePHP specific path information and our routing parameters. I felt this was the cleanest approach as I wasn't comfortable subclassing Diactoros\ServerRequest to add methods for these just yet.

Refs #6960

### CODE REVIEW

**Comment by jippi** [comment]
@markstory do you have a fully working implementation for PSR-7 somewhere? would love to see the performance implications on all this boilerplate we have add just to support a PSR :)

**Comment by ADmad** [comment]
@jippi https://github.com/markstory/cakephp-spekkoek

File: src/Http/ServerRequestFactory.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        array $server = null,
+        array $query = null,
+        array $body = null,
+        array $cookies = null,
+        array $files = null
+    ) {
+        $request = parent::fromGlobals($server, $query, $body, $cookies, $files);
+        list($base, $webroot) = static::getBase($request);
+        $request = $request->withAttribute('base', $base)
+            ->withAttribute('webroot', $webroot);
```
Comment (lorenzo) [inline src/Http/ServerRequestFactory.php]: Is it not possible to set multiple attributes at once?

Comment (markstory) [inline src/Http/ServerRequestFactory.php]: Sadly no.

**Comment by markstory** [comment]
@jippi The plugin @ADmad linked works, and is the code that is being incrementally added through smaller pull requests. I agree that it looks like a lot of code, but long term it will replace the following in CakePHP

* Cake\Network\Request
* Cake\Network\Response
* Cake\Routing\Dispatcher
* Cake\Routing\DispatcherFactory
* Cake\Routing\Filter\*

Which is a sizable chunk of code. From the testing I did there was a 2-5% decrease in requests per second. I attribute most of that to the translation layers we need for backwards compatibility.

**Comment by jippi** [comment]
@markstory alright ! :)

**Comment by lorenzo** [comment]
great job @markstory :)

**Pull request merged**